### PR TITLE
cocoa-cb: add support for macOS color space transformation (EDR/HDR)

### LINKED ISSUE #7341: utilise macOS HDR/EDR functionality
Edit: To those who wonder what EDR is, here's a brief explanation: [Apple’s “EDR” Brings High Dynamic Range to Non-HDR Displays](https://prolost.com/blog/edr)

I was testing Quicktime Player and mpv 0.31.0 the other day on my 2017 13" Macbook Pro running macOS Catalina 10.15.2. The material I've been using is an HDR video which is linked below.

https://mdt6.com/MDTTVTESTBETA082HEVC10BIT2020.mp4

At 00:00:08, the video displays three rectangles with different luminance. According to the [source](https://www.mdt6.com/?p=10647) (in Chinese unfortunately), the bottom one is 100 nits, the middle one is 350 nits while the top one being maximum of the screen.

With the default config `tone-mappinng=hable`, mpv is able to reveal the difference among three rectangles as below.

截屏2020-01-10下午7 29 38

However, Quicktime Player has significantly different behaviour. The rectangle on the top is brighter than mpv, even much brighter than a normal pure white picture (the screenshot below couldn't reproduce what's on the screen). If I manually turn up the brightness, the gap closes up. It seems that Quicktime Player is able to utilize more luminance of the screen than mpv, even though Macbook Pro only has an SDR one built-in.

截屏2020-01-10下午7 31 49

My expectation is to see the relatively "correct" luminance and colour within the capability of my screen. If I remember correctly, the config should be `tone-mapping=clip tone-mapping-param=1.0` and `target-peak` to be the brightness of my monitor. But with this setting, mpv still couldn't reach the brightness of Quicktime Player. Maybe the latter could automatically use the maximum brightness of the screen?

I'm wondering if they are treating luminance in fundamentally different ways. But I don't have a clue. Hope I could find an answer here.

### PR DESCRIPTION
**Pull request by Akemi**
up to date version of #8485, with some added new colour spaces. also only for cocoa-cb since it's not necessary for the macvk vulkan backend or rather would break it.

this will most likely be the last new feature for the cocoa-cb/libmpv/opengl backend from me. just wanted to add this already done feature that was resting for years.

it might come in handy for future backends though.

Fixes #7341

### CODE REVIEW

**Comment by Akemi** [comment]
on macOS 12 + Xcode 13.1 + SDK 12. can't find 2 of the 4 colour spaces introduced in 12.0, indicating an SDK problem
❌ https://developer.apple.com/documentation/coregraphics/cgcolorspace/4010583-itur_709_hlg
❌ https://developer.apple.com/documentation/coregraphics/cgcolorspace/3892091-itur_709_pq
✅ https://developer.apple.com/documentation/coregraphics/cgcolorspace/3867016-lineardisplayp3
✅ https://developer.apple.com/documentation/coregraphics/cgcolorspace/3867017-linearitur_2020
```
Message: Detected macOS sdk path: /Applications/Xcode_13.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.0.sdk
Message: Detected macOS SDK: 12.0

--------------------------------------------------------------------------------------

../video/out/cocoa_cb_common.swift:164:79: error: type 'CGColorSpace' has no member 'itur_709_HLG'
            case MAC_CSP_ITUR_709_HLG: return CGColorSpace(name: CGColorSpace.itur_709_HLG)
                                                                 ~~~~~~~~~~~~ ^~~~~~~~~~~~
../video/out/cocoa_cb_common.swift:165:78: error: type 'CGColorSpace' has no member 'itur_709_PQ'
            case MAC_CSP_ITUR_709_PQ: return CGColorSpace(name: CGColorSpace.itur_709_PQ)
                                                                ~~~~~~~~~~~~ ^~~~~~~~~~~
```

on macOS 12 + Xcode 14.2 + SDK 13.1. no building problems but following runtime error, indicating a general problem with those two new colour spaces and macOS 12 + CoreGraphic (framework version 12)
```
Message: Detected macOS sdk path: /Applications/Xcode_14.2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.1.sdk
Message: Detected macOS SDK: 13.1

--------------------------------------------------------------------------------------

dyld[8493]: Symbol not found: (_kCGColorSpaceITUR_709_HLG)
  Referenced from: '/Users/runner/work/mpv/mpv/build/mpv'
  Expected in: '/System/Library/Frameworks/CoreGraphics.framework/Versions/A/CoreGraphics'
./ci/build-macos.sh: line 25:  8493 Abort trap: 6           ./build/mpv -v --no-config
```

on macOS 13 + Xcode 14.2 + SDK 13.1 (same setup as second test), there are no problems regarding those colour spaces, indicating a problem with macOS 12 as a platform.

removing the two affected colour spaces on macOS 12 + Xcode 14.2 + SDK 13.1 shields no problems.

TLDR macOS 12 has a CoreGraphics framework issue, macOS SDK 12 is broken for those two colour spaces. the latest SDK available on macOS 12 is SDK 13, that fixes the build problems but not the runtime crash. is fixed with SDK 13 and macOS 13. to fix this, we can't just guard those two colour spaces with a build time and runtime check for macOS+SDK 13, since the symbol not found crash will still occur. for now i will just remove those two (undocumented) colour spaces.

i had similar issues with [colour spaces introduced in macOS 10.14.6](https://github.com/mpv-player/mpv/pull/8485/files#diff-5aca131483d9d070886f0f399f515ea4a1d58472e3fd9d06160707c2836f2b08R460-R473).

File: DOCS/man/options.rst [hunk 1]
```diff
@@ -6267,6 +6267,39 @@ them.
 
     macOS and cocoa-cb only.
 
+``--cocoa-cb-output-csp=<csp>``
+    This sets the color space of the layer to activate the macOS color
+    transformation. Depending on the color space used the system's EDR (HDR)
+    support will be activated. To get correct results, this needs to be set to
+    the color primaries/transfer characteristics of the video. It is recommended
```
Comment (kasper93) [inline DOCS/man/options.rst:6274]: > of the video

I think it is more correct to say that it has to have the same characteristics as VO target.

Comment (Akemi) [reply DOCS/man/options.rst:6274]: indeed true.

**Comment by KTamas** [comment]
oh my god I'm so excited about this one

File: osdep/mac/app_bridge.h [hunk 2]
```diff
@@ -35,6 +35,24 @@ enum {
...
+    MAC_CSP_ITUR_2020,
+    MAC_CSP_LINEAR_ITUR_2020,
+    MAC_CSP_ITUR_2100_HLG,
+    MAC_CSP_ITUR_2100_PQ,
+    MAC_CSP_ITUR_709,
+    MAC_CSP_SRGB,
+    MAC_CSP_LINEAR_SRGB,
+    MAC_CSP_GENERIC_RGB_LINEAR,
+    MAC_CSP_ADOBE_RGB1998,
+};
```
Comment (kasper93) [inline osdep/mac/app_bridge.h:54]: I don't like yet another CSP enum, but I guess it is unavoidable.

File: DOCS/man/options.rst [hunk 3]
```diff
@@ -6267,6 +6267,37 @@ them.
 
     macOS and cocoa-cb only.
 
+``--cocoa-cb-output-csp=<csp>``
+    This sets the color space of the layer to activate the macOS color
+    transformation. Depending on the color space used the system's EDR (HDR)
+    support will be activated. To get correct results, this needs to be set to
+    the color primaries/transfer characteristics of the VO target. It is recommended
+    to use this switch together with ``--target-trc``.
```
Comment (kasper93) [inline DOCS/man/options.rst:6275]: > ``--target-trc`` and ```--target-prim```

It seems that both are needed to properly replicate CSP.

File: DOCS/man/options.rst [hunk 4]
```diff
@@ -6267,6 +6267,37 @@ them.
...
+    :dcip3:              DCI P3 color space.
+    :itur_2020:          ITU BT.2020 color space.
+    :linear_itur_2020:   ITU BT.2020 color space and linear transfer function.
+    :itur_2100_hlg:      ITU BT.2100 and the Hybrid Log-Gamma (HLG) transfer function.
+    :itur_2100_pq:       ITU BT.2100 and the Perceptual Quantizer (PQ) transfer function.
+    :itur_709:           ITU BT.709 color space.
+    :srgb:               sRGB colorimetry and non-linear transfer function.
+    :linear_srgb:        Same as sRGB but linear transfer function.
+    :generic_rgb_linear: RGB and linear transfer function.
+    :adobe_rgb1998:      Adobe RGB (1998) color space.
```
Comment (kasper93) [inline DOCS/man/options.rst:6297]: I would unify the naming with other options like `target-prim`.

`display_p3` -> `display-p3`
`itur_709` -> `bt.709`

and so on.

Also I wonder does it really need to be set manually by user? Can't we infer those values based on `target-trc`, `target-prim` and `target-colorspace-hint` somehow?

File: meson.build [hunk 5]
```diff
@@ -1570,6 +1570,18 @@ if features['cocoa'] and features['swift']
...
+    macos_sdk_version.version_compare('>= 11'),
+    error_message: 'A macos sdk version >= 11 could not be found!',
+)
+features += {'macos-11-features': macos_11_features.allowed()}
+
+macos_12_features = get_option('macos-12-features').require(
+    macos_sdk_version.version_compare('>= 12'),
+    error_message: 'A macos sdk version >= 12 could not be found!',
+)
+features += {'macos-12-features': macos_12_features.allowed()}
```
[the remark this answers is not in the document]
Comment (kasper93) [reply meson.build:1583]: Also is this a best way to check SDK version, instead of actually definitions during build?

`__MAC_OS_X_VERSION_MAX_ALLOWED` or something like that.

[the remark this answers is not in the document]
Comment (Akemi) [reply meson.build:1583]: sure it does support building on lower SDKs. not every runtime check needs a build time check. the ones you mentioned are just a workaround for older version of macOS, the properties and variables used are available on all supported macOS versions.

to my knowledge this is the best way to check it. since SDK version is not the same as max allowed version. though i think re-evaluating this is something out of scope of this PR.

File: DOCS/man/options.rst (hunk 4 above)
Comment (Akemi) [reply DOCS/man/options.rst:6297]: renamed for consistency with other options:
```
display_p3 > display-p3
display_p3_hlg > display-p3-hlg
display_p3_pq > display-p3-pq
linear_display_p3 > display-p3-linear
dcip3 > dci-p3
itur_2020 > bt.2020
linear_itur_2020 > bt.2020-linear
itur_2100_hlg > bt.2100-hlg
itur_2100_pq > bt.2100-pq
itur_709 > bt.709
srgb > srgb
linear_srgb > srgb-linear
generic_rgb_linear > rgb-linear
adobe_rgb1998 > adobe
```

**Review by kasper93 — APPROVED** [review]
Not tested, but looks ok.

**Comment by KTamas** [comment]
Not sure this is working? Compiled this version just now. FIle is proven to be a proper HDR video when remuxed to MP4 for QuickTime, I use it as my test file for ages, It is streamed from the internet, my config is below. Works with IINA although IINA's HDR colors are completely wrong and way too dark compared to QuickTime.

```
macos-fs-animation-duration=0
ontop
ontop-level=window
window-scale=0.3
demuxer-max-bytes=4096MiB
demuxer-max-back-bytes=2048MiB
screenshot-format=jpg
screenshot-high-bit-depth=no
screenshot-directory=/Users/redacted/Desktop
save-position-on-quit
screenshot-tag-colorspace=yes
profile=fast
```

Screenshot 2024-05-01 at 6 15 42 PM

**Comment by Akemi** [comment]
this is for cocoa-cb/libmpv and not vo=gpu(-next). you are using the macvk backend.

>  also only for cocoa-cb since it's not necessary for the macvk vulkan backend or rather would break it.

**Comment by KTamas** [comment]
Right, my bad. Forced it to libmpv, however HDR is still not activated? Maybe I'm still doing something wrong. It's more apparent to the eyes but this is a screenshot with IINA on the left and MPV on the right, and even the screenshot shows the brightness difference.

image

**Comment by KTamas** [comment]
Right, my bad. Forced it to libmpv, however HDR is still not activated? Maybe I'm still doing something wrong. It's more apparent to the eyes but this is a screenshot with IINA on the left and MPV on the right, and even the screenshot shows the brightness difference.

image

**Comment by eko5624** [comment]
I want to test on my intel macbookpro 2017,  what's the right csp? How to write mpv.conf, just like:
```
vo=libmpv
cocoa-cb-output-csp=?
target-trc=?
target-prim=?
```

**Comment by Akemi** [comment]
i can't really tell. it depends on you hardware and display characteristics. though that is mentioned in the documentation.

also this PR is probably the wrong place for discussing this. you should rather open a discussion.

**Comment by KTamas** [comment]
fwiw I managed to get *some* HDR working by playing around with those parameters (it looked wrong, but it was HDR), so in that sense I can confirm it working. Appreciate the help.

**Comment by Akemi** [comment]
i tested it some more on my TV and it also seems to work properly when setup/configured correctly.

i might write up something in the wiki how to configure everything correctly in the future. though still need to test quite a lot still.
